casper/src/test/scala/coop/rchain/casper/HashSetCasperTest.scala

```diff
@@ -42,6 +42,31 @@ class HashSetCasperTest extends FlatSpec with Matchers {
   private val bonds   = createBonds(validators)
   private val genesis = buildGenesis(wallets, bonds, 0L)
 
+  it should "not replay when adding block from itself" in {
+    val nodes = HashSetCasperTestNode.network(validatorKeys.take(2), genesis)
+    val deployData = ProtoUtil.sourceDeploy(
+      """new s(`rho:io:stdout`) in { s!("Hello, World!") }""",
+      System.currentTimeMillis()
+    )
+    val Created(signedBlock) = nodes(0).casperEff
+      .deploy(deployData) *> nodes(0).casperEff.createBlock
+
+    nodes(0).casperEff.addBlock(signedBlock)
+    nodes(1).receive()
+
+    val received = nodes(1).casperEff.contains(signedBlock)
+    received should be(true)
+    nodes(0).logEff.infos
+      .count(it => it.contains("skip replay the deploys sent from itself")) should be(
+      1
+    )
```

Comment:
Before this commit, the "Hello, World!" will be printed three times in your console. Once for node0 deploy the data, and twice for both node1 and node0 replay the RSpace. And now, it will be only be printed twice. But the output is by the interpret. I have no idea to catch them. So I use the info log to prove.